{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Align import substitution_matrices\n",
    "from Bio import Entrez, SeqIO\n",
    "from Bio import Seq\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. Naloga"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "blosumMatrix = substitution_matrices.load(\"BLOSUM62\")\n",
    "def blosumFunction(x, y):\n",
    "    if x == \"-\":\n",
    "        return blosumMatrix[\"*\"][y]\n",
    "    elif y == \"-\":\n",
    "        return blosumMatrix[x][\"*\"]\n",
    "    else:\n",
    "        return blosumMatrix[x][y]\n",
    "def insertIntoString(string,char,position):\n",
    "    return string[:position] + char + string[position:len(string)]\n",
    "def global_allignment(seq1, seq2, scoring_function = blosumFunction):\n",
    "    lengthSeq1 = len(seq1)\n",
    "    lengthSeq2 = len(seq2)\n",
    "\n",
    "    optMatrix = [[0 for j in range(lengthSeq2 + 1)] for i in range(lengthSeq1 + 1)]\n",
    "    tracebackMatrix = [[('ins') if i == 0 else ('del') for j in range(lengthSeq2 + 1)] for i in range(lengthSeq1 + 1)]\n",
    "    tracebackMatrix[0][0] = ('')\n",
    "    for i in range(1, lengthSeq1 + 1):\n",
    "        optMatrix[i][0] = i * scoring_function(seq1[i - 1],\"-\") \n",
    "    for j in range(1, lengthSeq2 + 1):\n",
    "        optMatrix[0][j] = j * scoring_function(\"-\",seq2[j - 1]) \n",
    "\n",
    "    for i in range(1, lengthSeq1 + 1):\n",
    "        for j in range(1, lengthSeq2 + 1):\n",
    "            allign = optMatrix[i-1][j-1] + scoring_function(seq1[i -1],seq2[j-1])\n",
    "            delete = optMatrix[i - 1][j] + scoring_function(\"-\", seq2[j - 1])\n",
    "            insert = optMatrix[i][j-1] + scoring_function(seq1[i - 1],\"-\")\n",
    "            optMatrix[i][j] = max(allign, delete, insert)\n",
    "            \n",
    "            if optMatrix[i][j] == allign:\n",
    "                tracebackMatrix[i][j] = ('align')  \n",
    "            elif optMatrix[i][j] == insert:\n",
    "                tracebackMatrix[i][j] = ('ins') \n",
    "            elif optMatrix[i][j] == delete:\n",
    "                tracebackMatrix[i][j] = ('del')    \n",
    "\n",
    "\n",
    "    tracebackSteps = []\n",
    "    nextOperation = tracebackMatrix[lengthSeq1][lengthSeq2]\n",
    "    xCounter = lengthSeq1\n",
    "    yCounter = lengthSeq2\n",
    "    while True:\n",
    "        tracebackSteps.append(nextOperation)\n",
    "        if nextOperation== 'align':\n",
    "            xCounter -= 1\n",
    "            yCounter -= 1\n",
    "        elif nextOperation == 'ins':\n",
    "            yCounter -= 1\n",
    "        elif nextOperation == 'del':\n",
    "            xCounter -= 1\n",
    "        nextOperation = tracebackMatrix[xCounter][yCounter]\n",
    "        if nextOperation == (''):\n",
    "            break\n",
    "    tracebackSteps = tracebackSteps[::-1]\n",
    "\n",
    "    for i in range(0,len(tracebackSteps)):\n",
    "        action = tracebackSteps[i]\n",
    "        if action == \"allign\":\n",
    "            pass\n",
    "        elif action == \"ins\":\n",
    "            seq1 = insertIntoString(seq1,\"-\",i)\n",
    "        elif action == \"del\":\n",
    "            seq2 = insertIntoString(seq2,\"-\",i)\n",
    "\n",
    "    return (optMatrix[lengthSeq1][lengthSeq2], seq1, seq2)  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2. Naloga"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "Entrez.email = \"ls90321@student.uni-lj.si\"\n",
    "accession_codes = {\n",
    "    # 6 known human coronaviruses\n",
    "    \"Human-SARS\": \"NC_004718\",\n",
    "    \"Human-MERS\": \"NC_019843\",\n",
    "    \"Human-HCoV-OC43\": \"NC_006213\",\n",
    "    \"Human-HCoV-229E\": \"NC_002645\",\n",
    "    \"Human-HCoV-NL63\": \"NC_005831\",\n",
    "    \"Human-HCoV-HKU1\": \"NC_006577\",\n",
    "    \n",
    "    # Bat\n",
    "    \"Bat-CoV MOP1\": \"EU420138\",\n",
    "    \"Bat-CoV HKU8\": \"NC_010438\",\n",
    "    \"Bat-CoV HKU2\": \"NC_009988\",\n",
    "    \"Bat-CoV HKU5\": \"NC_009020\",\n",
    "    \"Bat-CoV RaTG13\": \"MN996532\",\n",
    "    \"Bat-CoV-ENT\": \"NC_003045\",\n",
    "    \n",
    "    # Other animals\n",
    "    \"Hedgehog-CoV 2012-174/GER/2012\": \"NC_039207\",\n",
    "    \"Pangolin-CoV MP789\": \"MT121216\",\n",
    "    \"Rabbit-CoV HKU14\": \"NC_017083\",\n",
    "    \"Duck-CoV isolate DK/GD/27/2014\": \"NC_048214\",\n",
    "    \"Feline infectious peritonitis virus\": \"NC_002306\",  # cat\n",
    "    \"Giraffe-CoV US/OH3/2003\": \"EF424623\",\n",
    "    \"Murine-CoV MHV/BHKR_lab/USA/icA59_L94P/2012\": \"KF268338\",  # mouse\n",
    "    \"Equine-CoV Obihiro12-2\": \"LC061274\",  # horse\n",
    "}\n",
    "genes = {}\n",
    "records = []\n",
    "for key in accession_codes.keys():\n",
    "    with Entrez.efetch(db=\"nucleotide\", id = accession_codes[key], rettype=\"gbwithparts\", retmode=\"FASTA\") as handle:\n",
    "        records.append(SeqIO.read(handle,\"genbank\"))\n",
    "    for record in records:\n",
    "        for gene in record.features:\n",
    "            if gene.type == 'CDS':\n",
    "                if 'gene' in gene.qualifiers.keys():\n",
    "                    if gene.qualifiers['gene'] == ['S'] or gene.qualifiers['product'] == ['spike protein']:\n",
    "                        genes[key] = gene\n",
    "                elif gene.qualifiers['product'] == ['spike protein']:\n",
    "                        genes[key] = gene\n",
    "\n",
    "    records = []\n",
    "\n",
    "covidSeq =\"\"\n",
    "firstLine = True\n",
    "with open(\"data/sars_cov_2.fa\") as f:\n",
    "    for line in f.readlines():\n",
    "        if not firstLine:\n",
    "            whitoutBackslash = line.strip(\"\\n\")\n",
    "            covidSeq += whitoutBackslash\n",
    "        else:\n",
    "            firstLine = False\n",
    "sarsCovGenomeSpike = str(Seq.Seq(covidSeq[21562:25384]).translate())\n",
    "allAnlignedScores = {}\n",
    "for key in genes.keys():\n",
    "    allAnlignedScores[key] = global_allignment(sarsCovGenomeSpike[0:len(sarsCovGenomeSpike)-1],genes[key].qualifiers['translation'][0])[0]\n",
    "\n",
    "\n",
    "allAnlignedScores\n",
    "\n",
    "three_closest_references = [\n",
    "    \"Bat-CoV RaTG13\",\n",
    "    \"Pangolin-CoV MP789\",\n",
    "    \"Human-SARS\",\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "3. Naloga"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "metadata": {},
   "outputs": [],
   "source": [
    "def local_alignment(seq1, seq2, scoring_function = blosumFunction ):\n",
    "    lengthSeq1 = len(seq1)\n",
    "    lengthSeq2 = len(seq2)\n",
    "\n",
    "    optMatrix = [[0 for j in range(lengthSeq2 + 1)] for i in range(lengthSeq1 + 1)]\n",
    "    \n",
    "    tracebackMatrix = [[('ins') if i == 0 else ('del') for j in range(lengthSeq2 + 1)] for i in range(lengthSeq1 + 1)]\n",
    "    tracebackMatrix[0][0] = ('')\n",
    "    for i in range(1, lengthSeq1 + 1):\n",
    "        optMatrix[i][0] = max(i * scoring_function(seq1[i - 1],\"-\"),0) \n",
    "    for j in range(1, lengthSeq2 + 1):\n",
    "        optMatrix[0][j] = max(j * scoring_function(\"-\",seq2[j - 1]),0) \n",
    "\n",
    "    maxScore = -1\n",
    "    maxScoreIndexS1 = 0\n",
    "    maxScoreIndexS2 = 0\n",
    "    for i in range(1, lengthSeq1 + 1):\n",
    "        for j in range(1, lengthSeq2 + 1):\n",
    "            allign = optMatrix[i-1][j-1] + scoring_function(seq1[i -1],seq2[j-1])\n",
    "            delete = optMatrix[i - 1][j] + scoring_function(\"-\", seq2[j - 1])\n",
    "            insert = optMatrix[i][j-1] + scoring_function(seq1[i - 1],\"-\")\n",
    "            elt = optMatrix[i][j] = max(allign, delete, insert, 0)\n",
    "            \n",
    "            if elt == allign or 0:\n",
    "                tracebackMatrix[i][j] = ('align')  \n",
    "            elif elt == insert:\n",
    "                tracebackMatrix[i][j] = ('ins') \n",
    "            elif elt == delete:\n",
    "                tracebackMatrix[i][j] = ('del')                 \n",
    "            if elt > maxScore:\n",
    "                maxScore = elt\n",
    "                maxScoreIndexS1 = i\n",
    "                maxScoreIndexS2 = j\n",
    "\n",
    "    tracebackSteps = []\n",
    "    nextOperation = tracebackMatrix[lengthSeq1][lengthSeq2]\n",
    "    xCounter = maxScoreIndexS1\n",
    "    yCounter = maxScoreIndexS2\n",
    "    while True:\n",
    "        tracebackSteps.append(nextOperation)\n",
    "        if nextOperation == 'align':\n",
    "            xCounter -= 1\n",
    "            yCounter -= 1\n",
    "        elif nextOperation == 'ins':\n",
    "            yCounter -= 1\n",
    "        elif nextOperation == 'del':\n",
    "            xCounter -= 1\n",
    "        nextOperation = tracebackMatrix[xCounter][yCounter]\n",
    "        if nextOperation == ('') or optMatrix[xCounter][yCounter] == 0:\n",
    "            break\n",
    "    tracebackSteps = tracebackSteps[::-1]\n",
    "\n",
    "    seq1 = seq1[xCounter:maxScoreIndexS1]\n",
    "    seq2 = seq2[yCounter:maxScoreIndexS2]\n",
    "    for i in range(0,len(tracebackSteps)):\n",
    "        action = tracebackSteps[i]\n",
    "        if action == \"allign\":\n",
    "            pass\n",
    "        elif action == \"ins\":\n",
    "            seq1 = insertIntoString(seq1,\"-\",i)\n",
    "        elif action == \"del\":\n",
    "            seq2 = insertIntoString(seq2,\"-\",i)\n",
    "\n",
    "\n",
    "    return (optMatrix[maxScoreIndexS1][maxScoreIndexS2], seq1, seq2)\n",
    "  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "4. Naloga"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "orfsAndTranslations = {}\n",
    "seq = str(Seq.Seq(covidSeq[11995:13483]).translate())\n",
    "seq = seq[0:len(seq) - 1]\n",
    "seq2 = str(Seq.Seq(covidSeq[26792:27191]).translate())\n",
    "seq2 = seq2[0:len(seq2) - 1]\n",
    "seq3 = str(Seq.Seq(covidSeq[23650:25384]).translate())\n",
    "seq3 = seq3[0:len(seq3) - 1]\n",
    "seq4 = str(Seq.Seq(covidSeq).reverse_complement()[len(covidSeq) - 667 : len(covidSeq) - 421].translate())\n",
    "seq4 = seq4[0:len(seq4) - 1]\n",
    "seq5 = str(Seq.Seq(covidSeq[9133:13483]).translate())\n",
    "seq5 = seq2[0:len(seq5) - 1]\n",
    "orfsAndTranslations[\"ORF-1\"] = seq\n",
    "orfsAndTranslations[\"ORF-2\"] = seq2\n",
    "orfsAndTranslations[\"ORF-3\"] = seq3\n",
    "orfsAndTranslations[\"ORF-4\"] = seq4\n",
    "orfsAndTranslations[\"ORF-5\"] = seq5  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "three_closest_references = [\n",
    "    \"Bat-CoV RaTG13\",\n",
    "    \"Pangolin-CoV MP789\",\n",
    "    \"Human-SARS\",\n",
    "]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [],
   "source": [
    "relativeData = {\"Bat-CoV RaTG13\": [],\"Pangolin-CoV MP789\":[],\"Human-SARS\":[]}\n",
    "records = []\n",
    "for key in three_closest_references:\n",
    "    with Entrez.efetch(db=\"nucleotide\", id = accession_codes[key], rettype=\"gbwithparts\", retmode=\"FASTA\") as handle:\n",
    "        records.append(SeqIO.read(handle,\"genbank\"))\n",
    "    for record in records:\n",
    "        for gene in record.features:\n",
    "            if gene.type == 'CDS':\n",
    "                relativeData[key].append(gene)\n",
    "    records = []\n",
    "    \n",
    "relativeFunctions = {\"Bat-CoV RaTG13\": [],\"Pangolin-CoV MP789\":[],\"Human-SARS\":[]}\n",
    "for key in relativeData.keys():\n",
    "    for gene in relativeData[key]:\n",
    "        data = gene\n",
    "        relativeFunctions[key].append([data.qualifiers[\"translation\"][0], data.qualifiers['product'][0]])\n",
    "closestRelatives = {}\n",
    "for key in orfsAndTranslations.keys():\n",
    "    maxAl = -1\n",
    "    for virusName in relativeFunctions.keys():\n",
    "        for gene in relativeFunctions[virusName]:\n",
    "            translatedVirus = gene[0]\n",
    "            virusFunction = gene[1]\n",
    "            aligment = local_alignment(translatedVirus, orfsAndTranslations[key])\n",
    "            score = aligment[0]\n",
    "            if score > maxAl:\n",
    "                maxAl = score\n",
    "                relativeName = virusName\n",
    "                relativeFunction = virusFunction\n",
    "                trLen = len(translatedVirus)\n",
    "    closestRelatives[key] = [relativeName,relativeFunction, maxAl, trLen]      "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "closestRelatives"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "6. Naloga"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "three_closest_references = [ \n",
    "    \"Human-MERS\",\n",
    "    \"Bat-CoV HKU5\",\n",
    "    \"Hedgehog-CoV 2012-174/GER/2012\",\n",
    "]\n",
    "relativeData = {\"Human-MERS\": [],\"Bat-CoV HKU5\":[],\"Hedgehog-CoV 2012-174/GER/2012\":[]}\n",
    "\n",
    "records = []\n",
    "for key in three_closest_references:\n",
    "    with Entrez.efetch(db=\"nucleotide\", id = accession_codes[key], rettype=\"gbwithparts\", retmode=\"FASTA\") as handle:\n",
    "        records.append(SeqIO.read(handle,\"genbank\"))\n",
    "\n",
    "    for record in records:\n",
    "        for gene in record.features:\n",
    "            if gene.type == 'CDS':\n",
    "                relativeData[key].append(gene)\n",
    "    records = []\n",
    "relativeFunctions = {\"Human-MERS\": [],\"Bat-CoV HKU5\":[],\"Hedgehog-CoV 2012-174/GER/2012\":[]}\n",
    "for key in relativeData.keys():\n",
    "    for gene in relativeData[key]:\n",
    "        data = gene\n",
    "        relativeFunctions[key].append([data.qualifiers[\"translation\"][0], data.qualifiers['product'][0]])\n",
    "closestRelatives = {}\n",
    "for key in orfsAndTranslations.keys():\n",
    "    maxAl = -1\n",
    "    for virusName in relativeFunctions.keys():\n",
    "        for gene in relativeFunctions[virusName]:\n",
    "            translatedVirus = gene[0]\n",
    "            virusFunction = gene[1]\n",
    "            aligment = local_alignment(translatedVirus, orfsAndTranslations[key])\n",
    "            score = aligment[0]\n",
    "            if score > maxAl:\n",
    "                maxAl = score\n",
    "                relativeName = virusName\n",
    "                relativeFunction = virusFunction\n",
    "                trLen = len(translatedVirus)\n",
    "    closestRelatives[key] = [relativeName,relativeFunction, maxAl, trLen]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ORF-1': ['Hedgehog-CoV 2012-174/GER/2012',\n",
       "  'ORF1ab polyprotein',\n",
       "  1419.0,\n",
       "  7150],\n",
       " 'ORF-2': ['Human-MERS', 'membrane protein', 322.0, 219],\n",
       " 'ORF-3': ['Human-MERS', 'spike protein', 1292.0, 1353],\n",
       " 'ORF-4': ['Human-MERS', '1AB polyprotein', 59.0, 7078],\n",
       " 'ORF-5': ['Human-MERS', 'membrane protein', 322.0, 219]}"
      ]
     },
     "execution_count": 183,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "closestRelatives"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.10.8 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "52634da84371cba311ea128a5ea7cdc41ff074b781779e754b270ff9f8153cee"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
